### ISSUE TITLE
SVG 2 spec section 13.5.7 (probably) incorrect for miter-clip (and arcs?)

### ISSUE THREAD

**Issue by skef** [w3c/svgwg#745]
(Note: I sent a version of this in email to @Tavmjong before discovering this was a better forum.)

This is in reference to sections [13.5.5](https://svgwg.org/svg2-draft/painting.html#LineJoin) and [13.5.7](https://svgwg.org/svg2-draft/painting.html#StrokeShape) of the current SVG 2 draft. It is also related to #545 

In 13.5.5 this is said about stroke-miterlimit:

> \<number\> The limit on the extent of a miter, miter-clip, or arcs line join as a multiple of the stroke-width value. A negative value for stroke-miterlimit is invalid and must be ignored.

>  Previous versions of the SVG specification also stated that values between 0 and 1 were in error, but this was not well implemented by user agent's CSS parsers. In practice, any miter join will exceed a miter limit between 0 and 1.

This implies that values of stroke-miterlimit less than 1 are not invalid in SVG 2.

In the sub-section of 13.5.7 that starts with "The line join shape for a given segment of a subpath ...", 11-13 of the spec say:

> 11. f stroke-linejoin is miter or miter-clip then the line join region is the union of bevel and the triangle formed from the three points P1, P2 and P3.

>  12. Let θ be the angle between A and B. If 1 / sin(θ / 2) ≤ stroke-miterlimit, then return the line join region.

>  13. If stroke-linejoin is miter-clip, then clip any part of the line join region that extends past the miter limit and return this region.

11 thereby specifies the meaning of "line join region" in this case.

13 is a bit vague here, as the description itself doesn't clarify the meaning of "extends past the miter limit". However, one can presume a clipping line parallel to P1-P2 at a distance f(miter limit) from P based on the earlier discussion.

However, suppose stroke-miterlimit is less than 1, perhaps significantly so. Or suppose it is a bit less than 1 and the join angle is small, putting the bevel distance close to stroke-width/2. In that case the spec, read strictly, requires little triangles at P1 and P2 extending beyond the clip line (because those are outside of the "line join region" as specified in 11).

Anyone reading the spec would likely assume that is not what is intended, but the correct interpretation could be either of:

* Never clip within the bevel triangle -- just return the bevel even if part of it is outside the specified clip range.
* Allow shorter clips but also clip the relevant portions outside the line join region (trimming off P1 and P2).

And indeed in #545 @AmeliaBR says "I didn't make any changes to the algorithms for the new miter types (miter-clip and arcs), since I thought it could make sense to actually want to clip those miters closer to the vertex than the normal stroke width."  So the state of the current text may be waiting on a substantive decision. 

However, as I've been working on a related implementation, I've come to doubt that anything but a bevel fall-back is likely to wind up in the spec. For one thing, unless the stroke shapes are lines, clipping "earlier" will change the tangent angles, potentially yielding a join that is quite different than expected. For another, if the angle is very acute there isn't much "benefit" to clipping back to the stroke-center, as you still counter-intuitively wind up with all of the length on the other side of the join. 

Given these two challenges it seems that any comprehensive solution to shorter stroke-miterlimts would need to based in part on the geometry near the join rather than just the angles and stroke width in isolation. And that would require a good deal of investigation and discussion. Therefore for SVG 2, at least, it's probably best to clarify that the bevel area is the fallback for all miter-clip joins with "short" stroke-miterlimits and to make the analogous changes for "arcs".

**Comment by Tavmjong** [comment]
I agree that for miter-clip, the clip should not be less than that which would produce a bevel. The situation where a clip would cut into the corners of the stroked path (without line joins) was not considered. I think it is highly unlikely that any graphics library would support clipping away the corners of stroked path. (Although we, Inkscape, do exactly that with our "Live Path Effect" implementation as we generate the path ourselves and don't rely on underlying graphics library in this case.)

For the arcs line join, the equivalent should be specified although this not as nice as the transition between clipping perpendicular to the circular arc (as specified) does not smoothly transition to a bevel but this should lead to a rather small discontinuity (if animated).

The key point is that clipping should not cause a visual jump in the rendering of the line join as the join angle is changed.

**Comment by Tavmjong** [comment]
A figure to help explain what is being discussed here.

![miter-clip](https://user-images.githubusercontent.com/8554441/67238505-90c50180-f44d-11e9-8827-897762cfd910.png)

**Comment by AmeliaBR** [comment]
In https://github.com/w3c/svgwg/issues/545#issuecomment-424112695, when we agreed to accept `stroke-miterlimit` values between 0 and 1, it was on the assumption that it wouldn't affect rendering.

**Comment by skef** [comment]
> In [#545 (comment)](https://github.com/w3c/svgwg/issues/545#issuecomment-424112695), when we agreed to accept `stroke-miterlimit` values between 0 and 1, it was on the assumption that it wouldn't affect rendering.

Hmm, but "affect rendering" in the sense that the outcome is the same for values between 0 and 1 that it would be for 1, or in the sense that the algorithm isn't more difficult to implement? The bevel line can still be shy of .9 (when the bevel triangle is equilateral, for example).

**Comment by skef** [comment]
@Tavmjong if you have a chance to answer I'm curious about an aspect of:

> Although we, Inkscape, do exactly that with our "Live Path Effect" implementation as we generate the path ourselves and don't rely on underlying graphics library in this case.

What does Inkscape do when the splines are curved so that clipping "earlier" changes the tangent vectors and therefore the miter angle?

**Comment by AmeliaBR** [comment]
>  "affect rendering" in the sense that the outcome is the same for values between 0 and 1 that it would be for 1

Yes, but specifically for the existing SVG 1 line join styles. (A `miter` join is always going to be equal to or greater than a limit of 1, so it would always get replaced with the bevel, even though the bevel is often much less than that limit.) We weren't considering the new cases in that discussion, as you discovered.

**Comment by skef** [comment]
@AmeliaBR Ah, right, sorry -- you were clarifying a historical fact and I read more than that into your statement.

**Comment by skef** [comment]
I:

1. Am not a person entrusted with writing specs and
2. Have not yet implemented the arcs part of my project. 

However, I've done enough research about recent thinking about these joins to know that solutions with continuity of result relative to continuity of parameter -- [often demonstrated with an animation](http://tavmjong.free.fr/SVG/LINEJOIN_STUDY/) -- tend to "win" over other options. So about this from earlier: 

> For the arcs line join, the equivalent should be specified although this not as nice as the transition between clipping perpendicular to the circular arc (as specified) does not smoothly transition to a bevel but this should lead to a rather small discontinuity (if animated).

Thinking about this, here is an illustration of a "full" arcs join with its bevel line: 

![a1](https://user-images.githubusercontent.com/6175836/67260596-0088bb80-f451-11e9-8d21-0c6005db56cf.png)

If the Miter Limit corresponds to a line where both points of the "clip line" intersect the curve to the left of the bevel line, that's easy: just use the bevel line. If both points of the clip line intersect the curve to the right of the bevel line that's also easy: use that line. The question is what to do when the intersection points are on different sides of the bevel line. 

If the best result is a continuous one the solution seems pretty straightforward: For the point "behind" (in this case to the left of) the bevel line, substitute its bevel point. So in the "virtual animation" the chosen line is the clip line until one end reaches the bevel line, then that point stays and the other point continues back (according to the clip line) until it reaches its bevel point, and lower values stay like that.

**Comment by Tavmjong** [comment]
I'm not actually sure what Inkscape is doing when the corners are clipped due to curved lines. I'll need to check.

I agree that the best result is as you've suggested. I was thinking the same thing.

**Comment by karlcow** [comment]
Closing as outdated. miter-clip and arcs removed from SVG 2 per #1065 (PR #1074).
